claims/access.go

```diff
@@ -29,17 +29,50 @@ type AccessRequest struct {
 
 	// For non-resource requests, this will be the requested URL path
 	Path string
+
+	// Folder is the parent folder of the requested resource
+	Folder string
+}
+
+type CheckResponse struct {
+	// Allowed is true if the request is allowed, false otherwise.
+	Allowed bool
 }
 
+type AccessChecker interface {
+	// Check checks whether the user can perform the given action for all requests
+	Check(ctx context.Context, id AuthInfo, req CheckRequest) (CheckResponse, error)
+}
+
+type ListRequest struct {
+	// The requested access verb.
+	// this includes get, list, watch, create, update, patch, delete, deletecollection, and proxy,
+	// or the lowercased HTTP verb associated with non-API requests (this includes get, put, post, patch, and delete)
+	Verb string
+
+	// API group (dashboards.grafana.app)
+	Group string
+
+	// ~Kind eg dashboards
+	Resource string
+
+	// tenant isolation
+	Namespace string
+
+	// Optional subresource
+	Subresource string
+}
+
+// TODO: Should the namespace be specified in the request instead.
+// I don't think we'll be able to Compile over multiple namespaces.
+// Checks access while iterating within a resource
+type ItemChecker func(namespace string, name string) bool
```

Comment:
```suggestion
type ItemChecker func(namespace, name, folder string) bool
```

each item will need a folder check 😢